ADF terraform extension (#397)
awslabs/aws-deployment-framework: docs/user-guide.md

# User Guide

>   - [Deployment Map](#deployment-map)

Author: revert to original toc

# User Guide

> - [User Guide](#user-guide)

Author: removed from toc

**Overview**

ADF support the deployment of terraform code to multiple accounts and regions through terraform pipelines.
The module consists of four build stages defined in the following CodeBuild build specification:

- `buildspec.yml`: install the version of terraform specified in the pipeline configuration
- `tf_scan.yml`: (optional) scans for vulnerabilities in the terraform code using the terrascan application. If vulnerabilities are found, it will fail and block further execution in the pipeline. It is recommended to enable this step in all ADF terraform pipelines.
- `tf_plan.yml`: get the list of accounts from the organization and run a terraform plan.
- `tf_apply.yml`: get the list of accounts from the organization and run a terraform plan and apply.

An optional approval step could be added between plan and apply as shown in the pipeline definition below.

Please look into the [sample-terraform](../samples/sample-terraform) pipeline for more details in the setup and integration.

**Parameters**

> - `TERRAFORM_VERSION`: the terraform version used to deploy the resource.

Author: added comment related to buildspec configuration

- `TARGET_ACCOUNTS`: comma separated list of target accounts.
- `TARGET_OUS`: comma separated list of target leaf OUs (parent OUs are supported).

> - `REGIONS`: comma separated list of target regions. If this parameter is empty, the main ADF region is used.

Author: added management account id to the list

**Deployment procedure**

1. Add a sample-terraform pipeline in ADF `deployment-map.yml` as in the example:

```yaml
- name: sample-terraform
  default_providers:
    source:
      provider: codecommit
      properties:
        account_id: 111111111111 # source account id
    build:
      provider: codebuild
    deploy:
      provider: codebuild
      properties:
        image: "STANDARD_5_0"
        environment_variables:
          TARGET_ACCOUNTS: 111111111111,222222222222 # target accounts
          TARGET_OUS: /core/infrastructure,/sandbox # target OUs
          MANAGEMENT_ACCOUNT_ID: 333333333333 # management account / billing account
          REGIONS: eu-west-1 # target regions. Add a comma separated list to define multiple regions e.g. eu-west-1,us-east-1
  targets:
    - name: terraform-scan # optional
      properties:
        spec_filename: tf_scan.yml # terraform scan
    - name: terraform-plan
      properties:
        spec_filename: tf_plan.yml # terraform plan
    - approval # manual approval
    - name: terraform-apply
      properties:
        spec_filename: tf_apply.yml # terraform apply
```

2. Add the project name in params/global.yml file

> 3. Add terraform code to the `tf` folder. Do not make changes to `backend.tf` file and `main.tf`.

Author: added details

```

By passing in the Repository name *(repository)* we are overriding the **name** property which normally is the name of our associated repository. This will tie both of these pipelines to the single *sample-vpc* repository on the *111111111111* AWS Account.

### Terraform pipeline

> **Pre-requisites**

Reviewer: ```suggestion
#### Prerequisites
```

To enable ADF Terraform extension the following steps are required:
- Rename file `example-global-iam.yml` to `global-iam.yml` in the following path `aws-deployment-framework-bootstrap/adf-bootstrap/` and ensure the CloudFormation resources `ADFTerraformRole` and `ADFTerraformPolicy` are no longer commented out.
- Rename file `example-global-iam.yml` to `global-iam.yml` in the following path `aws-deployment-framework-bootstrap/adf-bootstrap/deployment` (please note: `deployment` at the end) and ensure the CloudFormation resources `ADFTerraformRole` and `ADFTerraformPolicy` are no longer commented out.

Important note: `ADFTerraformPolicy` IAM policy is a sample. This policies should **NOT** be used for purposes other than testing. You should scope this policy depending on what you would like to deploy using Terraform within the selected Organizational Units.

> **Overview**

Reviewer: ```suggestion
#### Overview
```

ADF support the deployment of Terraform code to multiple accounts and regions through Terraform pipelines.
The module consists of four build stages defined in the following CodeBuild build specification:

- `buildspec.yml`: install the version of Terraform specified in the pipeline configuration.
- `tf_scan.yml`: (optional) scans for vulnerabilities in the terraform code using the terrascan application. If vulnerabilities are found, it will fail and block further execution in the pipeline. It is recommended to enable this step in all ADF terraform pipelines.
- `tf_plan.yml`: get the list of accounts from the organization and run a terraform plan.
- `tf_apply.yml`: get the list of accounts from the organization and run a terraform plan and apply.

An optional approval step could be added between plan and apply as shown in the pipeline definition below.

Please look into the [sample-terraform](../samples/sample-terraform) pipeline for more details in the setup and integration.

> **Parameters**

Reviewer: ```suggestion
#### Parameters
```

- `TERRAFORM_VERSION`: the Terraform version used to deploy the resource. This parameter must be defined in the `buildspec.yml` file of the repository.
- `TARGET_ACCOUNTS`: comma separated list of target accounts.
- `TARGET_OUS`: comma separated list of target leaf OUs (parent OUs are supported).
- `REGIONS`: comma separated list of target regions. If this parameter is empty, the main ADF region is used.
- `MANAGEMENT_ACCOUNT_ID`: id of the AWS Organizations management account.

> **Deployment procedure**

Reviewer: ```suggestion
#### Deployment procedure
```

1. Add a sample-terraform pipeline in ADF `deployment-map.yml` as in the example:

```yaml
- name: sample-terraform
  default_providers:
    source:
      provider: codecommit
      properties:
        account_id: 111111111111 # source account id
    build:
      provider: codebuild
    deploy:
      provider: codebuild
      properties:
        image: "STANDARD_5_0"
        environment_variables:
          TARGET_ACCOUNTS: 111111111111,222222222222 # target accounts
          TARGET_OUS: /core/infrastructure,/sandbox # target OUs
          MANAGEMENT_ACCOUNT_ID: 333333333333 # management account / billing account
          REGIONS: eu-west-1 # target regions. Add a comma separated list to define multiple regions e.g. eu-west-1,us-east-1
  targets:
    - name: terraform-scan # optional
      properties:
        spec_filename: tf_scan.yml # Terraform scan
    - name: terraform-plan
      properties:
        spec_filename: tf_plan.yml # Terraform plan
    - approval # manual approval
    - name: terraform-apply
      properties:
        spec_filename: tf_apply.yml # Terraform apply
```

2. Add the project name in `params/global.yml` file.

> 3. Add Terraform code to the `tf` folder. Do not make changes to `backend.tf` file and `main.tf` which contain the definition of the remote state file location, Terraform provider definition. Any change to these files could affect the standard functionalities of this module.

Reviewer: ```suggestion
3. Add Terraform code to the `tf` folder. **Please note**: Do not make changes to `backend.tf` file and `main.tf` in the root folder of the sample. These contain the definition of the remote state file location and the Terraform provider definition. Any change to these files could disrupt the standard functionalities of this module.
```

5. Define the desired `TERRAFORM_VERSION` in the `buildspec.yml` file as shown in the sample-terraform example. ADF supports Terraform version v0.13.0 and later.
6. Push to your Terraform ADF repository, for example the sample-terraform one.
7. Pipeline contains a manual approval step between Terraform plan and Terraform apply. Confirm to proceed.

Terraform state files are stored in the regional S3 buckets in the deployment account. One state file per account/region/module is created.

e.g.

- Project name: sample-tf-module
- Target accounts: 111111111111, 222222222222
- Target regions: eu-west-1 (main ADF region), us-east-1

The following state files are created:

- 111111111111 main region (eu-west-1) adf-global-base-deployment-pipelinebucketxyz/sample-tf-module/111111111111.tfstate
- 111111111111 secondary region (us-east-1) adf-regional-base-deploy-deploymentframeworkregio-jsm/sample-tf-module/111111111111.tfstate
- 222222222222 main region (eu-west-1) adf-global-base-deployment-pipelinebucketxyz/sample-tf-module/222222222222.tfstate
- 222222222222 secondary region (us-east-1) adf-regional-base-deploy-deploymentframeworkregio-jsm/sample-tf-module/222222222222.tfstate

> A DynamoDB table is created to manage the lock of the state file. It is deployed in every ADF regions named `adf_locktable`. Please note: usage of this locking table is charged on the deployment account.

Reviewer: ```suggestion
A DynamoDB table is created to manage the lock of the state file. It is deployed in every ADF regions named `adf_locktable`. **Please note**: usage of this locking table is charged on the deployment account.
```

### Terraform pipeline

**Pre-requisites**
To enable ADF Terraform extension the following steps are required:
- Rename file example-global-iam.yml to global-iam.yml in the following path aws-deployment-framework-bootstrap/adf-bootstrap/ and comment out the Cloudformation resources ADFTerraformRole and ADFTerraformPolicy
- Rename file example-global-iam.yml to global-iam.yml in the following path aws-deployment-framework-bootstrap/adf-bootstrap/deployment and comment out the Cloudformation resources ADFTerraformRole and ADFTerraformPolicy

NOTE: ADFTerraformPolicy IAM policy is a sample. This policies should NOT be used for purposes other than testing. You should scope this policy depending on what you would like to deploy within certain Organizational Units.

**Overview**

ADF support the deployment of terraform code to multiple accounts and regions through terraform pipelines.
The module consists of four build stages defined in the following file:

> - `buildspec.yml`: install the version of terraform specified in the pipeline configuration

Author: I do agree. Added to buildspec.yml and updated the documentation
